### PR TITLE
Use generator to get rules

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | https://github.com/yiisoft/validator/issues/261

### CODE REVIEW

**Comment by samdark** [comment]
@xepozz is it still valid? If so, would you please resolve conflicts?

File: src/RulesProvider/AttributesRulesProvider.php [hunk 1]
```diff
@@ -14,9 +16,9 @@
 final class AttributesRulesProvider implements RulesProviderInterface
 {
     /**
-     * @var array<RuleInterface[]>|null
+     * @var Generator|null
      */
-    private ?array $rules = null;
+    private Generator|null $rules = null;
```
Comment (arogachev) [inline src/RulesProvider/AttributesRulesProvider.php:21]:
```suggestion
    private ?Generator $rules = null;
```

File: src/RulesProvider/AttributesRulesProvider.php [hunk 2]
```diff
@@ -28,38 +30,58 @@ public function __construct(
...
-        return $this->rules;
+
+        yield from $this->rules;
     }
 
     /**
-     * @return array<RuleInterface[]>
+     * @throws ReflectionException
+     *
+     * @return Generator
```
Comment (arogachev) [inline src/RulesProvider/AttributesRulesProvider.php:49]: Return types are different in PHPDoc / type hint.

File: src/Rule/NestedHandler.php [hunk 3]
```diff
@@ -103,7 +104,7 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
             }
 
             $validatedValue = ArrayHelper::getValueByPath($data, $valuePath);
-            $rules = is_array($rules) ? $rules : [$rules];
+            $rules = is_array($rules) || $rules instanceof Generator ? $rules : [$rules];
```
Comment (arogachev) [inline src/Rule/NestedHandler.php:107]:
```suggestion
            $rules = is_iterable($rules) ? $rules : [$rules];
```

File: tests/DataSet/PHP81/ObjectDataSet81Test.php [hunk 4]
```diff
@@ -28,7 +30,9 @@ public function testCollectRules(object $object, array $expectedRules): void
     {
         $dataSet = new ObjectDataSet($object);
 
-        $this->assertEquals($expectedRules, $dataSet->getRules());
+        $arrayRules = $this->toNestedArray($dataSet->getRules());
```
Comment (arogachev) [inline tests/DataSet/PHP81/ObjectDataSet81Test.php:33]: Grammatically these would be better options - `$arrayOfRules` / `$rulesArray` / `$rulesArr`. Or maybe `$actualRules`?

File: tests/DataSet/PHP81/ObjectDataSet81Test.php [hunk 5]
```diff
@@ -5,6 +5,8 @@
 namespace Yiisoft\Validator\Tests\DataSet\PHP81;
 
 use PHPUnit\Framework\TestCase;
+use Traversable;
+use Yiisoft\Validator\Attribute\Embedded;
```
Comment (arogachev) [inline tests/DataSet/PHP81/ObjectDataSet81Test.php:9]: Broken import.

**Review by arogachev — CHANGES_REQUESTED** [review]
:+1:

File: src/RulesProvider/AttributesRulesProvider.php (hunk 1 above)
Comment (arogachev) [reply src/RulesProvider/AttributesRulesProvider.php:21]: Also `null` can be omitted as it matches default value.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
@xepozz suggested not to merge it yet.

TODO:

- [ ] Benchmarks

File: src/RulesProvider/AttributesRulesProvider.php (hunk 2 above)
Comment (arogachev) [reply src/RulesProvider/AttributesRulesProvider.php:46]: Still actual.

File: src/RulesProvider/AttributesRulesProvider.php [hunk 6]
```diff
@@ -4,19 +4,18 @@
...
 use Yiisoft\Validator\RuleInterface;
 use Yiisoft\Validator\RulesProviderInterface;
 
 final class AttributesRulesProvider implements RulesProviderInterface
 {
-    /**
-     * @var array<RuleInterface[]>|null
-     */
-    private ?array $rules = null;
+    private Generator|iterable|null $rules = null;
```
Comment (arogachev) [inline src/RulesProvider/AttributesRulesProvider.php:18]:
```suggestion
    private Generator|iterable|null $rules;
```

**Comment by arogachev** [comment]
Need to merge with master and resolve conflicts. After that there are 4 failing tests:

> `InvalidArgumentException` : Each rule should be an instance of `Yiisoft\Validator\RuleInterface`.

File: src/RulesProvider/AttributesRulesProvider.php (hunk 6 above)
Comment (xepozz) [reply src/RulesProvider/AttributesRulesProvider.php:20]: It's impossible to used typed property without default value out of the constructor.

**Comment by xepozz** [comment]
> Need to merge with master and resolve conflicts. After that there are 4 failing tests with the same error:
> 
> > `InvalidArgumentException` : Each rule should be an instance of `Yiisoft\Validator\RuleInterface`.

Done with https://github.com/yiisoft/validator/pull/265/commits/9a467d1a9c8c107b32dacb9b72558d9ccc5d0701

File: src/RulesProvider/AttributesRulesProvider.php (hunk 2 above)
Comment (xepozz) [reply src/RulesProvider/AttributesRulesProvider.php]: Done with https://github.com/yiisoft/validator/pull/265/commits/0cf1c8811d0fd089221e11d398d55632802a55f1

**Pull request merged**